### PR TITLE
Regularize how we handle "impossible" conditions in internal APIs in C#.

### PR DESCRIPTION
**Pull request by gafter**
In a switch's default clause that is unexpected, for example, we should

```cs
    throw ExceptionUtilities.UnexpectedValue(value);
```
Long ago this was a standard but we've drifted from it.

In particular, these should not be Debug.Assert(false), because we end up continuing execution from an untested and poorly understood state in production (i.e. the production and development code paths differ). There is no value in making the exception conditional on Debug, as Debug.Assert does, except perhaps during a brief time period in which the code is under development. Unfortunately in many cases we never went back and turned them into "permanent" assertions.

@jaredpar @AlekseyTs @VSadov @agocke @khyperia

### CODE REVIEW

File: docs/compilers/Design/Unexpected Conditions.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
...
+- If you write a switch statement that is intended to be exhaustive of the possibilities, add a default branch with `throw ExceptionUtilities.UnexpectedValue(switchExpression);`.
+- Similarly, if you have a sequence of `if-then-else if` statements that is intended to be exhaustive of the possibilities, add a final "impossible" else branch with `throw ExceptionUtilities.Unreachable;` or, if is convenient to provide interesting, easy to obtain data for the diagnostic, use `ExceptionUtilities.UnexpectedValue`. Do this also for other situations where the code reaches a point that is "impossible".
+- Validation of preconditions in public APIs should be done with plain code. Report a diagnostic if one exists for the situation (e.g. a syntax error), or throw an appropriate exception such as `NullArgumentException` or `ArgumentException`.
```
Comment (jaredpar) [inline docs/compilers/Design/Unexpected Conditions.md]: Should be `ArgumentNullException` 

https://msdn.microsoft.com/en-us/library/system.argumentnullexception(v=vs.110).aspx

**Comment by gafter** [comment]
@jaredpar @AlekseyTs @VSadov @agocke @khyperia All comments addressed. Please review. I'd like to pull this today if there are no issues.

**Comment by gafter** [comment]
@dotnet-bot test this please

**Comment by gafter** [comment]
@jaredpar @AlekseyTs @VSadov @agocke @khyperia All comments addressed. Please review. I'd like to pull this today if there are no issues.

**Comment by agocke** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 2]
```diff
@@ -325,8 +325,8 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
                                 case ConversionKind.ImplicitUserDefined:
                                     return false;
                                 default:
-                                    Debug.Assert(false, "Unhandled conversion kind in reordering logic");
-                                    return false;
+                                    // Unhandled conversion kind in reordering logic
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: I think returning false might be acceptable in this case. It is ok to say that something is not safe to reorder.

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: The author went to some trouble to call out each conversion kind. If we return false here then the previous cases aren't needed.

I'd prefer to make functional changes like that in a separate changeset.

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: Ok, makes sense.
